Duplicati/Library/Main/Operation/Backup/BackendUploader.cs

```diff
@@ -85,98 +120,331 @@ public static Task Run(Common.BackendHandler backend, Options options, Common.Da
                 }
-            });
-                                
+            }
+
+            return false;
+        }
+
+        private void ResetBackend(Exception ex, Worker worker)
+        {
+            try
+            {
+                worker.Backend?.Dispose();
+            }
+            catch (Exception dex)
+            {
+                Logging.Log.WriteWarningMessage(LOGTAG, "BackendDisposeError", dex, "Failed to dispose backend instance: {0}", ex?.Message);
+            }
+            finally
+            {
+                worker.Backend = null;
+            }
+        }
+
+        private async Task RenameFileAfterErrorAsync(FileEntryItem item)
+        {
+            var p = VolumeBase.ParseFilename(item.RemoteFilename);
+            var guid = VolumeWriterBase.GenerateGuid();
+            var time = p.Time.Ticks == 0 ? p.Time : p.Time.AddSeconds(1);
+            var newname = VolumeBase.GenerateFilename(p.FileType, p.Prefix, guid, time, p.CompressionModule, p.EncryptionModule);
+            var oldname = item.RemoteFilename;
+
+            await m_stats.SendEventAsync(item.Operation, BackendEventType.Rename, oldname, item.Size);
+            await m_stats.SendEventAsync(item.Operation, BackendEventType.Rename, newname, item.Size);
+            Logging.Log.WriteInformationMessage(LOGTAG, "RenameRemoteTargetFile", "Renaming \"{0}\" to \"{1}\"", oldname, newname);
+            await m_database.RenameRemoteFileAsync(oldname, newname);
+            item.RemoteFilename = newname;
+        }
+
+        private async Task DoPut(FileEntryItem item, IBackend backend, CancellationToken cancelToken)
+        {
+            if (cancelToken.IsCancellationRequested)
+                return;
+
+            if (item.TrackedInDb)
+                await m_database.UpdateRemoteVolumeAsync(item.RemoteFilename, RemoteVolumeState.Uploading, item.Size, item.Hash);
+
+            if (m_options.Dryrun)
+            {
+                Logging.Log.WriteDryrunMessage(LOGTAG, "WouldUploadVolume", "Would upload volume: {0}, size: {1}", item.RemoteFilename, Library.Utility.Utility.FormatSizeString(new FileInfo(item.LocalFilename).Length));
+                item.DeleteLocalFile();
+                return;
+            }
+
+            await m_database.LogRemoteOperationAsync("put", item.RemoteFilename, JsonConvert.SerializeObject(new { Size = item.Size, Hash = item.Hash }));
+            await m_stats.SendEventAsync(BackendActionType.Put, BackendEventType.Started, item.RemoteFilename, item.Size, updateProgress: false);
+            m_progressUpdater.StartFileProgress(item.RemoteFilename, item.Size);
+
+            var begin = DateTime.Now;
+
+            if (!m_options.DisableStreamingTransfers && backend is IStreamingBackend streamingBackend)
+            {
+                using (var fs = File.OpenRead(item.LocalFilename))
+                using (var ts = new ThrottledStream(fs, m_initialUploadThrottleSpeed, 0))

```

Comment:
Did behaviour change here? Before a download limit was set, which is removed by setting the third argument to 0?
Suppose I have no upload limit, do not want my download bandwith to exceed 100kB/s, and I am uploading with 20MB/s and in order to do so, I download 500kB/s. Does this imply that I will violate the download 100kB/s limit?
Was this respected previously?

I am not saying either is correct or wrong, but if this is the case, behaviour changed. If I understand correctly, to maintain the old behaviour, you'd have to also start doing transfer speed bookkeeping in the other "download" direction in FileProgressThrottler?

Reply:
You are correct, that would be a change in behavior.
When I was playing with the code on my own I set that to 0 and thought, why do I need to limit the download speed when I'm uploading! I kind of glossed over it after that, but I agree that needs to be put back.

Reply:
I looked into this and I don't think this matters after all. When uploading data we of course upload it, but we don't download anything until the verify step. And that step uses different code to do the downloading.

Additionally the `ProgressReportingStream` only has one value that it tracks for progress so it cannot currently differentiate between downloads and uploads.

All _that_ said I did add in the code to handle throttling downloads but it will not currently be used so I don't know if it is worth it to push that code. What do you think?

Reply:
I'm not very familiar with this part of the code, but it sounds like it at least deserves of a comment explaining why the download throttle value is what it is.

Reply:
That makes sense, I'll add a comment as to why there is no download throttle value.

Reply:
I meant that even if you don't actively download files (like for verification), still packets are sent to be able to upload. It's two ways, you can't have 0 bytes in one direction without affecting the other side. My guess was the current code might take care of that, but who knows. I would not push unused code.

Reply:
Some data is downloaded when we upload in terms of TCP ACK packets for instance. Or if an uploaded packet is lost a retry will occur but that happens below the level of Duplicati so we aren't even aware of it, let alone able to measure it.